### ISSUE TITLE
Should error when invalid project options are passed

### ISSUE THREAD

**Issue by TingPing** [mesonbuild/meson#932]
Many times in the past I've seen packages passing invalid arguments that were either removed, typo-ed, etc.  It would be helpful to be very explicit when this happens.

**Comment by keszybz** [comment]
Copying my comment from #2095:

If the user makes a typo, it'd be best for them if meson threw an error. If the user tries to configure with an option that was removed, it'd probably be better for them to just get a warning. If they try to use an option which was renamed, an error would be more useful. So I don't know if an error or a warning is more appropriate, but IMHO it should at least print a big warning in red.

FTR, autotools says:
```console
$ ../configure --enable-foobar=false
configure: WARNING: unrecognized options: --enable-foobar
...
```

**Comment by rhd** [comment]
This just bit me today. I forgot to change a meson option on the CI server and meson was happy and just took the default, thus my CI was producing bad builds.

meson really should error out if a `Dfoo=bar` but is not defined in `meson_options.txt`.

**Comment by kanavin** [comment]
Yes please. This is something that Yocto project would appreciate a lot, and lack of such warnings or errors just caused us unnecessary regressions and extra work:
http://lists.openembedded.org/pipermail/openembedded-core/2018-January/146679.html
http://lists.openembedded.org/pipermail/openembedded-core/2018-January/146680.html

**Comment by nirbheek** [comment]
Hum, I am surprised this is not implemented yet.

**Comment by lazka** [comment]
In case breaking existing setups is a concern you could make it an error only in case `os.isatty(sys.stdout.fileno())`

**Comment by dcbaker** [comment]
I worked on this, and added warnings with a deprecation period. It's probably been long enough now we could just turn those warnings into errors.

**Comment by rohieb** [comment]
@dcbaker can you reference any commit or PR? I'd like to know if I need to cherry-pick anything to get this… 🙂

**Comment by eli-schwartz** [comment]
This issue was duplicated by #7288 and fixed by #8974

In meson 0.60.0 this became an error.
